### PR TITLE
Made private/internal fields/properties readonly

### PR DESCRIPTION
**Pull request by bkoelman**
Updated member signatures throughout the codebase to use `readonly` where possible. RoslynLight.sln builds in VS2015 CTP6 and BuildAndTest.proj runs all tests without errors.

### CODE REVIEW

**Comment by bkoelman** [comment]
@amcasey, @gafter, @AlekseyTs, @VSadov, @agocke @mavasani @srivatsn @heejaechang @Pilchie @pharring 

would some of you please review?

**Comment by pharring** [comment]
:+1:
I didn't examine every line in detail, but the sample I looked at looked fine.

Note that the compiler doesn't warn if you call a methoed passing a readonly field by ref, so watch out for that being a subtle semantic difference. I don't know if you checked for that when you made this change.

**Comment by bkoelman** [comment]
@pharring Did not know that. RoslynLight.sln contains 8692 lines with "ref" and 7771 with "readonly", so that's a bit too much to review manually. Should we make an analyzer to detect those cases and run it with and without this change, then compare?

**Comment by gafter** [comment]
This looks right to me. Anybody else?

**Comment by Pilchie** [comment]
LGTM

**Comment by pharring** [comment]
@bkoelman Actually, I might be wrong. I recall being caught out by passing a readonly field by ref in the past, but perhaps it was an earlier compiler bug. @Pilchie does this ring a bell with you?

**Comment by gafter** [comment]
@pharring I think that's right in VB. You can pass a readonly property or field by ref, but it won't be changed by the call. You can even pass a literal!

```vb
Option Strict On

Module Module1
    ReadOnly Property P As Boolean
        Get
            Return False
        End Get
    End Property
    ReadOnly F As Boolean = False
    Sub SetTrue(ByRef p As Boolean)
        p = True
    End Sub

    Sub Main()
        SetTrue(P)
        SetTrue(F)
        SetTrue(False)
        Console.WriteLine(P)
        Console.WriteLine(F)
        Console.WriteLine(False)
    End Sub

End Module
```

This program outputs

```none
False
False
False
```

**Comment by VSadov** [comment]
you cannot pass readonly fields as ref/out directly.
However compiler allows invoking mutating members on readonly receivers. 

Typical pattern is:

```C#

      private readonly SomeStructEnumerator e;

      {
           . . .
           e.MoveNext();    // <-- this call will be dome on a copy of e, this.e will not be updated.
  
      }
      
```

**Comment by bkoelman** [comment]
@pharring I'm playing with an analyzer I just wrote to detect readonly argument passed to ref parameter in C#. It gave me [CS0192](https://msdn.microsoft.com/en-us/library/sa4td06k%28v=vs.90%29.aspx). So this is not allowed, except for constructors.

**Comment by pharring** [comment]
@bkoelman VSadov's example is what I was remembering. When you make a struct field readonly, you'd better be sure that you don't call mutating methods on that field. The trouble is, there's no simple way to detect a mutating method call.

**Comment by bkoelman** [comment]
So I ran my analyzer. Without my changes, RoslynLight.sln contains 39 occurences in C# where a ctor ref/out parameter is invoked with a readonly field. After applying my changes, it finds 98 occurences.

So, how to proceed? Would you like me to remove all readonly fields (C#/VB) I added in structs? Does that eliminate the potential issues?

**Comment by VSadov** [comment]
I think the answer is yes, just to be sure. It is not extremely common for struct members to be mutating, but missing just couple cases could result in bad behavior.
in addition to that, even if member is not mutating, there could be a perf penalty for making a copy of the struct field before making a call (it is an IL requirement that 'this' is passed as a ref to struct methods, and there is no way to get a ref off a readonly struct field directly, so compiler must copy)

In short, it is safer to apply "make readonly" fix only to fields that are reference types or primitive numeric types.

**Comment by bkoelman** [comment]
@VSadov First I wrote an [analyzer](https://github.com/bkoelman/CodeAnalyzers/tree/master/ReadonlyFieldsInStructs/ReadonlyFieldsInStructs) that reports all readonly fields which are typed as structs, struct arrays or are generic with a struct constraint (see also unit tests for [CS](https://github.com/bkoelman/CodeAnalyzers/blob/master/ReadonlyFieldsInStructs/ReadonlyFieldsInStructs.Test/AnalyzerTestsCSharp.cs) and [VB](https://github.com/bkoelman/CodeAnalyzers/blob/master/ReadonlyFieldsInStructs/ReadonlyFieldsInStructs.Test/AnalyzerTestsVisualBasic.cs)).
In Master 1269 occurences were reported, in my branch 1314. So I reverted these 45 changes, merged, resolved conflicts and rebased the pull request.

**Comment by VSadov** [comment]
I think the right filtering logic for the "make readonly" would be - fields that are definitely known to be reference types or primitives (int, uint, long, float . . . )
That would, for example, include arrays regardless of element type, but exclude interfaces or generic types not constrained to 'class' a reference type.

**Comment by pharring** [comment]
@VSadov why exclude interfaces?

**Comment by VSadov** [comment]
Actually interfaces are ok, Generic types constrained only to interfaces are not ok

**Comment by VSadov** [comment]
If the filtering is done via symbol APIs, I believe it should be possible to just check ITypeSymbol.IsReferenceType    it should know about details with generics.

**Comment by gafter** [comment]
I think this pull request is good to go. Support for primitives and some generics can be done in a later PR.

**Comment by VSadov** [comment]
I am slightly concerned that there could be some fields that are generic without 'struct' constraint - unconstrained or constrained to interfaces only and now marked as readonly by this change. 

We may not have such code, but it would be nice to be sure if we want to allow only reference fields to be automatically switched to readonly.

**Comment by gafter** [comment]
I'm going to see if any tests fail. Given the coding patterns we use in the compiler, I'd be very surprised if there is an issue.

**Comment by bkoelman** [comment]
I'm a bit surprised this pull request got merged in already, as I'm still working on updating my scanner based on the latest comments. But I will keep working on that anyway.

@VSadov Which exactly are primitive types? I'm not sure which list to pick from http://source.roslyn.codeplex.com/#Microsoft.CodeAnalysis.CSharp/Symbols/SpecialTypeExtensions.cs - I tend to believe `IsIntrinsicType(...)`

@pharring I don't understand why to allow interface types in making readonly. Structs can implement interfaces, so they may not be safe, right?

**Comment by gafter** [comment]
If I jumped the gun I can back this out for further work. @VSadov would you prefer I do that?

**Comment by pharring** [comment]
@bkoelman yes, a struct can implement an interface but, if you have a field that is an interface obtained from a struct, then the struct has to be boxed (and the interface refers to the boxed copy). No further copying is necessary in order to call (potentially mutating) methods on the readonly interface.

**Comment by VSadov** [comment]
@gafter  if we are getting a follow up change, no need to go back and forth. The chances are fairly low that we have made some thing readonly unintentionally.

@bkoelman , re: interfaces.  when field is an interface it is a reference type. Even if it is assigned a struct, it will be in a boxed form. That is different from a field of generic type T constrained to an interface. In such case it can be literally substituted to a struct type.
It is ok to turn field like " IEnumerator<int> f " to be readonly, it is not ok to make readonly a field like "T f", where T: IEnumerator<int>.  In the second case, we will get into troubles with MoveNext if T is a struct, which is relatively common (see List<T>.Enumerator).
It is unlikely to see this kind of fields in Roslyn, but it would be nice to be sure.

IsIntrinsicType does not map one-to-one with CLR primitive types, but for this purpose it seems good enough.

**Comment by bkoelman** [comment]
Clear explanation, I did not consider boxing, thanks!

I'm now using `ITypeSymbol.IsReferenceType`, which works consistently except here: On C# code:

```C#
using System;

namespace N
{
    public class O
    {
        public class C<T> where T : IDisposable
        {
            private readonly T f;
        }
    }
}
```

it returns `true`, but on VB code:

```VB
Option Strict On

Namespace N
	Public Class O
		Public Class C(Of T As IDisposable)
			Private ReadOnly f As T
		End Class
	End Class
End Namespace
```

it returns `false`. Should I create an issue for this?

**Comment by VSadov** [comment]
Are you calling IsRefernceType on T ?
in such case this would seem to be a bug in C#.  T is not a reference type.

**Comment by VSadov** [comment]
Just to be clear - both    IsReferenceType and   IsValueType should return false, since we cannot tell from constraints.

**Comment by bkoelman** [comment]
Yes I'm calling it on `T`. But I reversed `true` and `false`, so the bug is in VB.

IsReferenceType, IsPrimitiveType, IsValueType
VB: T, F, F
CS: F, F, F

**Comment by VSadov** [comment]
I looked through implementation on VB side it seems same as in C#.
One possible explanation is that VB sample is missing "Imports System" and as such "IDisposable" binds to an error symbol which looks more like a class than a struct.

**Comment by bkoelman** [comment]
@VSadov Very sharp, you are right!

Probably due to http://source.roslyn.codeplex.com/#Microsoft.CodeAnalysis.CSharp/Symbols/ErrorTypeSymbol.cs,89

So, not a bug after all.

**Comment by bkoelman** [comment]
Hi all, good news!

After updating [my scanner](https://github.com/bkoelman/CodeAnalyzers/tree/master/ReadonlyFieldsInStructs) (based on discussions here) it reported no new files to be rolled back. So all should be good now.

**Comment by gafter** [comment]
@bkoelman Awesome, thanks!
